garage/tf/regressors/categorical_mlp_regressor_with_model.py

```diff
@@ -0,0 +1,255 @@
+"""A regressor based on MLP with Normalized Inputs."""
+from dowel import tabular
+import numpy as np
+import tensorflow as tf
+
+from garage.tf.distributions import Categorical
+from garage.tf.misc import tensor_utils
+from garage.tf.models import NormalizedInputMLPModel
+from garage.tf.optimizers import ConjugateGradientOptimizer, LbfgsOptimizer
+from garage.tf.regressors import StochasticRegressor2
+
+
+class CategoricalMLPRegressorWithModel(StochasticRegressor2):
+    """
+    CategoricalMLPRegressor with garage.tf.models.NormalizedInputMLPModel.
+
+    A class for performing regression (or classification, really) by fitting
+    a Categorical distribution to the outputs. Assumes that the output will
+    always be a one hot vector
+
+    Args:
+        input_shape (tuple[int]): Input shape of the training data.
+        output_dim (int): Output dimension of the model.
+        name (str): Model name, also the variable scope.
+        hidden_sizes (list[int]): Output dimension of dense layer(s) for
+            the MLP for the network. For example, (32, 32) means the MLP
+            consists of two hidden layers, each with 32 hidden units.
+        hidden_nonlinearity (callable): Activation function for intermediate
+            dense layer(s). It should return a tf.Tensor. Set it to
+            None to maintain a tanh activation.
+        hidden_w_init (callable): Initializer function for the weight
+            of intermediate dense layer(s). The function should return a
+            tf.Tensor. Default is Glorot uniform initializer.
+        hidden_b_init (callable): Initializer function for the bias
+            of intermediate dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        output_nonlinearity (callable): Activation function for output dense
+            layer. It should return a tf.Tensor. Set it to None to
+            maintain a softmax activation.
+        output_w_init (callable): Initializer function for the weight
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is Glorot uniform initializer.
+        output_b_init (callable): Initializer function for the bias
+            of output dense layer(s). The function should return a
+            tf.Tensor. Default is zero initializer.
+        optimizer (tf.Optimizer): Optimizer for minimizing the negative
+            log-likelihood.
+        optimizer_args (dict): Arguments for the optimizer. Default is None,
+            which means no arguments.
+        tr_optimizer (tf.Optimizer): Optimizer for trust region approximation.
```

Comment:
is this actually a tf.Optimizer or is it a garage optimizer?

Reply:
I think it should be a garage.tf.Optimizer? Is there a tf second order optimizer?

Reply:
It should be `garage.tf.Optimizer`, will change that.

Reply:
Can you also update the same thing for `ContinuousMLPRegressorWithModel` and `GaussianMLPRegressorWithModel`?

Reply:
Sure